### PR TITLE
 Refactor type & bounds parsing thoroughly

### LINKED ISSUE #67146: Parser accepts negative lifetime bounds and interprets as positive bound
The following is accepted by the parser:
```rust
fn foo<T: !'static>() {}
```
and the `!` is just silently dropped thus interpreting the constraint as `T: 'static`.

Regression introduced in https://github.com/rust-lang/rust/pull/57364 on Feb 24.
cc @estebank @hdhoang

I discovered this while working on a refactoring of `parser/ty.rs`.
I will fix the issue as part of that refactoring.

### PR DESCRIPTION
**Pull request by Centril**
PR is based on https://github.com/rust-lang/rust/pull/67131 with first one from this PR being ` extract parse_ty_tuple_or_parens`.

Also fixes #67146.

r? @estebank

### CODE REVIEW

File: src/librustc_parse/parser/ty.rs [hunk 1]
```diff
@@ -461,6 +452,27 @@ impl<'a> Parser<'a> {
...
+        &mut self,
+        lo: Span,
+        inner_lo: Span,
+        has_parens: bool,
+        question: Option<Span>,
+    ) -> PResult<'a, GenericBound> {
+        self.error_opt_out_lifetime(question);
+        let bound = GenericBound::Outlives(self.expect_lifetime());
+        if has_parens {
+            // FIXME(Centril): Consider not erroring here and accepting `('lt)` instead,
```
Comment (petrochenkov) [inline src/librustc_parse/parser/ty.rs]: `('lt)` is entirely useless and it's some PITA to parse it in the first trait object position like `type A = `('lt) + OtherBounds`, so it isn't accepted anywhere.

**Comment by petrochenkov** [comment]
I still feel like half of these changes (extracting small functions used once) make code readability worse, but you are the primary person working with this code now, so feel free to adjust it to your personal tastes basically.

**Comment by Centril** [comment]
https://github.com/rust-lang/rust/pull/67131 has merged & I've rebased this PR so it should be ready for review now.

File: src/librustc_parse/parser/ty.rs [hunk 2]
```diff
@@ -319,130 +275,241 @@ impl<'a> Parser<'a> {
...
-                err.span_suggestion_hidden(
-                    bound_list,
-                    &format!("remove the trait bound{}", pluralize!(negative_bounds_len)),
-                    new_bound_list,
-                    Applicability::MachineApplicable,
-                );
+                new_bound_list = new_bound_list.replacen(" +", ":", 1);
             }
-            err.emit();
+            err.span_suggestion_hidden(
```
Comment (estebank) [inline src/librustc_parse/parser/ty.rs]: Change to tool_only_span_suggestion (the text is redundant, but the tooltip will appear in VSCode and rustfix will see it).

File: src/librustc_parse/parser/ty.rs [hunk 3]
```diff
@@ -319,130 +275,241 @@ impl<'a> Parser<'a> {
...
+    }
+
+    fn error_negative_bounds(
+        &self,
+        colon_span: Option<Span>,
+        bounds: &[GenericBound],
+        negative_bounds: Vec<Span>,
+    ) {
+        let negative_bounds_len = negative_bounds.len();
+        let last_span = *negative_bounds.last().unwrap();
```
Comment (estebank) [inline src/librustc_parse/parser/ty.rs]: Are you _really_ sure this will _never_ fail ^_^.

Make it `expect` if you are.

**Review by estebank — APPROVED** [review]
LGTM

r=me after addressing nitpicks (including ignoring them)

File: src/librustc_parse/parser/ty.rs (hunk 3 above)
Comment (Centril) [reply src/librustc_parse/parser/ty.rs]: By `if !negative_bounds.is_empty() {` (https://github.com/rust-lang/rust/pull/67148/files#diff-4b628dd7c6e6204b8378ca5a05472f5fR368) we know that `negative_bounds` is non-empty and so `.last().unwrap()` cannot panic. I'll change to `.expect(...)` for clarity tho.

File: src/librustc_parse/parser/ty.rs (hunk 2 above)
Comment (Centril) [reply src/librustc_parse/parser/ty.rs]: Done.

**Comment by Centril** [comment]
Addressed the comments in the two commits above, @bors r=estebank

**Pull request merged**